### PR TITLE
blockstore: archival file as a circular buffer

### PR DESCRIPTION
**Pull request by emwang-jumptrading**
test_archive_block tests:
- when blocks written greatly exceed block_idx
- when blocks written greatly exceed file size 
- rereading from an archive file and continuing to update the archive
- rereading from an archive file that has overwritten blocks / wrapped around blocks

### CODE REVIEW

File: src/flamenco/runtime/fd_blockstore.c [hunk 1]
```diff
@@ -58,7 +58,10 @@ fd_blockstore_new( void * shmem,
   fd_rwseq_new( &blockstore->lock );
   FD_COMPILER_MFENCE();
 
-  blockstore->off  = 0;
+  blockstore->off         = FD_BLOCKSTORE_ARCHIVE_START;
+  blockstore->fd_size_max = FD_BLOCKSTORE_ARCHIVE_MIN_SIZE;
+  blockstore->lrw_slot    = FD_SLOT_NULL;
+  blockstore->mrw_slot    = FD_SLOT_NULL;
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.c:64]: i would track these in a separate struct `fd_blockstore_archiver` and base the API around that

File: src/flamenco/runtime/fd_blockstore.c [hunk 2]
```diff
@@ -190,6 +193,52 @@ static inline void check_read_write_err( int err ) {
   }
 }
 
+bool 
+fd_blockstore_invalid_archive( fd_blockstore_t * blockstore, fd_blockstore_archive_metadata_t * metadata ) {
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.c:197]: same idea as above eg. `fd_blockstore_archiver_verify( fd_blockstore_archiver_t * )`

File: src/flamenco/runtime/fd_blockstore.c [hunk 3]
```diff
@@ -190,6 +193,52 @@ static inline void check_read_write_err( int err ) {
...
+         || ( metadata->lrw_off < FD_BLOCKSTORE_ARCHIVE_START ) 
+         || ( metadata->mrw_end < FD_BLOCKSTORE_ARCHIVE_START )
+         || ( metadata->fd_size_max != blockstore->fd_size_max ); // should be initialized same as archive file
+}
+
+#define check_read_err_safe( cond, msg )               \
+  if( FD_UNLIKELY( cond ) ) {                     \
+    FD_LOG_WARNING(( "[%s] %s", __func__, msg )); \
+    return FD_BLOCKSTORE_ERR_SLOT_MISSING;        \
+  }
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.c]: kind of a funky macro here to pass in the `msg` like that

you can just use string concatenation `##` c preprocessor syntax and so `[%s] ## msg` that

File: src/flamenco/runtime/fd_blockstore.h [hunk 4]
```diff
@@ -318,7 +322,10 @@ struct __attribute__((aligned(FD_BLOCKSTORE_ALIGN))) fd_blockstore {
 
   /* Persistence */
 
-  ulong off; /* current offset in the archival file */
+  ulong fd_size_max; /* maximum size of the archival file */
+  ulong off;         /* current offset in the archival file after most recent write */
+  ulong lrw_slot;    /* least recently written slot. maintains the file circular buffer */
+  ulong mrw_slot;    /* most recently written slot. maintains the file circular buffer */
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.h:326]: track in `fd_blockstore_archiver`

File: src/flamenco/runtime/fd_blockstore.h [hunk 5]
```diff
@@ -718,28 +725,52 @@ struct fd_blockstore_ser {
...
 /* Restores a block and block map entry from fd at given offset. As this used by
-   rpcserver, it must return an error code instead of throwing an error on failure */
+   rpcserver, it must return an error code instead of throwing an error on failure.
+   Modifies read_off so that it points to the end of the most recent read. If we read
+   to EOF, it will circle back around to read_off = FD_BLOCKSTORE_ARCHIVE_START */
 int
 fd_blockstore_block_meta_restore( fd_blockstore_t * blockstore,
                                   int fd,
-                                  fd_block_idx_t * block_idx_entry,
+                                  ulong * read_off,
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.h:756]: API can just take `fd_blockstore_archiver_t *` all tracked together

File: src/flamenco/runtime/fd_blockstore.c [hunk 6]
```diff
@@ -190,7 +193,55 @@ static inline void check_read_write_err( int err ) {
...
+                                 ulong dst_sz, 
+                                 ulong * rsz, 
+                                 ulong * read_off ) {
+  check_read_err_safe( lseek( fd, (long)*read_off, SEEK_SET ) == -1, 
+                       "failed to seek to read offset" );
+  
+  ulong remaining_sz = blockstore->fd_size_max - *read_off;
+  if ( remaining_sz < dst_sz ) {
+    int err = fd_io_read( fd, dst, remaining_sz, remaining_sz, rsz );
+    check_read_err_safe( err, "failed to read file near end" );
```
Comment (asiegel-jt) [inline src/flamenco/runtime/fd_blockstore.c:223]: if remaining_sz != *rsz, it is also an error

File: src/flamenco/runtime/fd_blockstore.c [hunk 7]
```diff
@@ -548,56 +621,206 @@ fd_blockstore_buffered_shreds_remove( fd_blockstore_t * blockstore, ulong slot )
...
+                                   ulong write_off ) {
+
+  if ( FD_UNLIKELY( lseek( fd, (long)write_off, SEEK_SET ) == -1 ) ) {
+    FD_LOG_ERR(( "[%s] failed to seek to offset %lu", __func__, write_off ));
+  }
+  ulong wsz;
+  ulong remaining_sz = blockstore->fd_size_max - write_off;
+  if ( remaining_sz < src_sz ) {
+    int err = fd_io_write( fd, src, remaining_sz, remaining_sz, &wsz );
+    check_read_write_err( err );
```
Comment (asiegel-jt) [inline src/flamenco/runtime/fd_blockstore.c:640]: check that remaining_sz == wsz

File: src/flamenco/runtime/fd_blockstore.c [hunk 8]
```diff
@@ -548,56 +621,206 @@ fd_blockstore_buffered_shreds_remove( fd_blockstore_t * blockstore, ulong slot )
...
+  ulong wsz;
+  int err = fd_io_write( fd, &metadata, sizeof(metadata), sizeof(metadata), &wsz );
+  check_read_write_err( err );
+}
+
+/* Clears any to be overwritten blocks in the archive from the index and updates lrw slot */
+static ulong
+fd_blockstore_lrw_archive_clear( fd_blockstore_t * blockstore, ulong slot, ulong wsz, ulong write_off ) {
+  fd_block_idx_t * block_idx = fd_blockstore_block_idx( blockstore );
+  fd_block_idx_t * lrw_block_index = fd_block_idx_query(block_idx, blockstore->lrw_slot, NULL);
```
Comment (asiegel-jt) [inline src/flamenco/runtime/fd_blockstore.c:691]: what if lrw_slot is null?

Comment (emwang-jumptrading) [reply src/flamenco/runtime/fd_blockstore.c:691]: then lrw_block_index would be null, which is handled -- but I'll change the check to be on the slot itself so its more clear

File: src/flamenco/runtime/fd_blockstore.c (hunk 6 above)
Comment (emwang-jumptrading) [reply src/flamenco/runtime/fd_blockstore.c:223]: I believe [fd_io_read already returns](https://github.com/firedancer-io/firedancer/blob/4aa0d7b17c40e433875aa7fb65a7efc71507ff2f/src/util/io/fd_io.h#L251) an errcode < 0 if rsz < remaining_sz?

File: src/flamenco/runtime/fd_blockstore.c (hunk 7 above)
Comment (emwang-jumptrading) [reply src/flamenco/runtime/fd_blockstore.c:640]: ditto to ^^

File: src/flamenco/runtime/fd_blockstore.c (hunk 6 above)
Comment (asiegel-jt) [reply src/flamenco/runtime/fd_blockstore.c:223]: if the file is shorter than expected or we run out of disk space, read/write will return a smaller value

File: src/flamenco/runtime/fd_blockstore.h [hunk 9]
```diff
@@ -718,28 +723,50 @@ struct fd_blockstore_ser {
...
 
-/* Archives a block and block map entry to fd. If fd is -1, no write is attempted */
+/* fd_blockstore_archive_metadata outlines the format of metadata
+   at the start of an archive file - needed so that archive
+   files can be read back on initialization. */
+
+struct fd_blockstore_archive_metadata {
+  ulong fd_size_max;
+  ulong lrw_off;
+  ulong mrw_end;
```
Comment (asiegel-jt) [inline src/flamenco/runtime/fd_blockstore.h:733]: add a couple of ulong's for future usage

**Review by asiegel-jt — APPROVED** [review]

File: src/flamenco/runtime/fd_blockstore.h (hunk 9 above)
Comment (lidatong) [reply src/flamenco/runtime/fd_blockstore.h:733]: we can just version the file

File: src/flamenco/runtime/fd_blockstore.c (hunk 6 above)
Comment (emwang-jumptrading) [reply src/flamenco/runtime/fd_blockstore.c:238]: read/write returns a smaller value in the rsz, but the err code returned is negative, which is checked in `check_read_err_safe` -- is this what you are referring to?

File: src/flamenco/runtime/fd_blockstore.h (hunk 9 above)
Comment (asiegel-jt) [reply src/flamenco/runtime/fd_blockstore.h]: then include a magic number

**Review by asiegel-jt — APPROVED** [review]

**Pull request merged**